comp/syntheticstestscheduler/impl/scheduler.go

```diff
@@ -0,0 +1,236 @@
+		if err := json.Unmarshal(rawConfig.Config, &syntheticsTestCfg); err != nil {
+			s.log.Warnf("skipping invalid Synthetics Test update %s: %v", configPath, err)
+			applyStateCallback(configPath, state.ApplyStatus{
+				State: state.ApplyStateError,
+				Error: "error unmarshalling payload",
+			})
+			continue
+		}
+
+		newConfig[syntheticsTestCfg.PublicID] = syntheticsTestCfg
+		s.log.Debugf("Processed config %s: %v", configPath, syntheticsTestCfg)
+
+		applyStateCallback(configPath, state.ApplyStatus{State: state.ApplyStateAcknowledged})
+	}
+
+	s.updateRunningState(newConfig)
+	s.runningConfig = newConfig
+
+	if err := s.persistConfig(); err != nil {
+		s.log.Warnf("unable to persist synthetics test config: %v", err)
+		// TODO: what path should I provide if it is global to the config
+		applyStateCallback("TODO", state.ApplyStatus{
+			State: state.ApplyStateError,
+			Error: "unable to persist synthetics test config",
+		})
+	}
+}
+
+// updateRunningState synchronizes in-memory runtime state with a new configuration.
+func (s *SyntheticsTestScheduler) updateRunningState(newConfig map[string]common.SyntheticsTestConfig) {
+	s.state.mu.Lock()
+	defer s.state.mu.Unlock()
+
+	seen := map[string]struct{}{}
+	for _, newTestConfig := range newConfig {
+		pubID := newTestConfig.PublicID
+		seen[pubID] = struct{}{}
+		current, exists := s.state.tests[pubID]
+		if !exists {
+			s.state.tests[pubID] = &runningTestState{
+				cfg:     newTestConfig,
+				lastRun: time.Time{},
+				nextRun: s.TimeNowFn().Add(time.Duration(newTestConfig.Interval) * time.Second),
+			}
+		} else {
+			current.cfg = newTestConfig
+		}
+	}
+
+	for pubID := range s.state.tests {
+		if _, exists := seen[pubID]; !exists {
+			delete(s.state.tests, pubID)
+		}
+	}
+}
+
+// persistConfig writes the runningConfig to persistent cache.
+func (s *SyntheticsTestScheduler) persistConfig() error {
+	cacheValue, err := json.Marshal(s.runningConfig)
+	if err != nil {
+		return err
+	}
+	if err := persistentcache.Write(cacheKey, string(cacheValue)); err != nil {
+		return s.log.Errorf("couldn't write cache: %s", err)
+	}
+	return nil
+}
+
+// retrieveConfig reads scheduler config from persistent cache and initializes state.
+func (s *SyntheticsTestScheduler) retrieveConfig() error {
+	cacheValue, err := persistentcache.Read(cacheKey)
+	if err != nil {
+		return s.log.Errorf("couldn't read from cache: %s", err)
+	}
+	if len(cacheValue) == 0 || cacheValue == "{}" {
+		return nil
+	}
+
+	var cfg map[string]common.SyntheticsTestConfig
+	if err := json.Unmarshal([]byte(cacheValue), &cfg); err != nil {
+		return err
+	}
+	s.runningConfig = cfg
+
+	s.state.mu.Lock()
+	defer s.state.mu.Unlock()
+
+	s.state.tests = make(map[string]*runningTestState, len(cfg))
+	now := s.TimeNowFn()
+	for id, testCfg := range cfg {
+		s.state.tests[id] = &runningTestState{
+			cfg:     testCfg,
+			lastRun: time.Time{},
+			nextRun: now.Add(time.Duration(testCfg.Interval) * time.Second),
```

Comment:
Hum, aren't we supposed to persist in cache the nextRun and re-load it?

I mean, for example, if a test should run once every week.

- Let's say at t0 the test is run, we persist the data
- and t0+3 days, the agent restarts, at this point, we should still ideally run the test at t0+1week right? which seems not the case with the current impl.